### PR TITLE
sm64ex: init at unstable-2020-06-19

### PR DESCRIPTION
**Pull request by IvarWithoutBones**
###### Motivation for this change
Adds [sm64pc](https://github.com/sm64pc/sm64ex), a PC port of Super Mario 64 based off of decompilation. 

The assets and such (the port itself extracts these) are copyrighted by Nintendo. Hence the need for the user to supply it themselves.

Note that i added an empty list of `makeFlags`, this is so you can easily add them as an end user for subjective patches. You can find two of them in the projects readme. For example, one of them includes a (subjectively) better camera, which the end user can now easily enable using overwrites.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by emilazy** [comment]
> I've licensed this as public domain, as the decompilation itself is that.

There's no explicit license statement or copyright waiver in the repository that I can see, and as the [reverse-engineered source code](https://github.com/n64decomp/sm64) it's based on is [modified decompiler output](https://arstechnica.com/gaming/2020/05/beyond-emulation-the-massive-effort-to-reverse-engineer-n64-source-code/) that compiles to an executable bit-identical to the original, it seems almost certain to me that it's legally a derivative work of the original game. As cool as this project is, I think the package must at least be marked as unfree; it would be a massive liability for Hydra to distribute binaries of it.

**Comment by IvarWithoutBones** [comment]
> There's no explicit license statement or copyright waiver in the repository that I can see

From what I've been told on their Discord server there's none provided, i assumed it was public domain. image

However, when i just looked into this, it seems the [AUR PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=sm64pc-git#n9) says it's MIT. Not sure what to do here. 

> It seems like the `extract_assets.py` script supports multiple versions of the ROM, whereas fixing the hash in the package means you can only use one specific release.

Indeed, i wanted to support this, and as you said not use a direct hash, but i wasn't sure how to go about it. I've based the user input of the rom on [mathematica](https://github.com/NixOS/nixpkgs/blob/3c4096f18bde58a6d7f2edc4defea2765e0b496f/pkgs/applications/science/math/mathematica/9.nix#L28), as im not aware of a better way to do this.

**Comment by emilazy** [comment]
"No license provided" isn't the same as "public domain", it's actually the opposite. Things are "all rights reserved" by default; to release something into the public domain requires an explicit waiver of automatic copyright protection (and even then it's only valid in some jurisdictions).

Even if there was a clear statement that all the copyrightable work of the reverse-engineers was MIT-licensed, it still wouldn't mean that the package as a whole can be used and redistributed under those terms, since that would require *all* the copyright holders of the work to MIT-license their contributions, and Nintendo have an extremely plausible copyright claim to the reverse-engineered source code it's based on (it compiles to an executable bit-identical to theirs, and was almost certainly originally machine-decompiled from it to begin with). In that case, the licensing situation could be documented like:

```nix
{
  meta = with stdenv.lib; {
    license = [
      # Original game code
      licenses.unfree
      # Work by reverse-engineers
      licenses.mit
    ];
  };
}
```

But unfortunately, there's no getting away from the `unfree` here, short of clean-room reverse engineering the entire game from scratch without any use of a decompiler.

> Indeed, i wanted to support this, and as you said not use a direct hash, but i wasn't sure how to go about it. I've based the user input of the rom on [mathematica](https://github.com/NixOS/nixpkgs/blob/3c4096f18bde58a6d7f2edc4defea2765e0b496f/pkgs/applications/science/math/mathematica/9.nix#L28), as im not aware of a better way to do this.

I think the best approach here is probably to take a `baseRom` argument that replaces the `requireFile` expression if it's not `null` so that people can override it if they want. The existing default seems fine to me, especially if it's used in existing packages for proprietary software (I hadn't seen `requireFile` used before).

**Comment by IvarWithoutBones** [comment]
> "No license provided" isn't the same as "public domain", it's actually the opposite. Things are "all rights reserved" by default; to release something into the public domain requires an explicit waiver of automatic copyright protection (and even then it's only valid in some jurisdictions).

I wasn't aware of that. Interesting. Since I don't actually see any mention of MIT in the project, should we just go with `unfree`?

> I think the best approach here is probably to take a `baseRom` argument that replaces the `requireFile` expression if it's not `null` so that people can override it if they want. The existing default seems fine to me, especially if it's used in existing packages for proprietary software (I hadn't seen `requireFile` used before).

Apologies, but I'm quite a beginner at nix. how would i go about doing that? I can't seem to figure it out with without encountering infinite recursion

**Comment by MetaDark** [comment]
@IvarWithoutBones Thanks for packaging this! I tried this out and it works perfectly.

Regarding licensing, sm64pc depends on n64-fast32-engine, which has a license stating that the source can be redistributed, but not the binary. This would definitely have to be marked as unfree.

Regarding @emilazy's comment, I think she meant that you could default `baseRom` to null and use an `if` statement to only specify the default when the argument is null:

```nix
{ stdenv, fetchFromGitHub, requireFile, python3, pkg-config
, audiofile, SDL2, hexdump, makeFlags ? []
, baseRom ? null
}:

stdenv.mkDerivation rec {
  ...
  baseRom =
    if baseRom != null then baseRom
    else
      requireFile {
        name = "baserom.us.z64";
            message = ''
          This nix expression requires that baserom.us.z64 is
          already part of the store. To get this file you can dump your US Super Mario 64 cartridge's contents
          and add it to the nix store with nix-store --add-fixed sha256 <FILE>.
        '';
        sha256 = "17ce077343c6133f8c9f2d6d6d9a4ab62c8cd2aa57c40aea1f490b4c8bb21d91";
      };
  ...
```
But that would lead to infinite recursion unless you named the argument something different. You could directly set the default instead to avoid name clashes:
```nix
{ stdenv, fetchFromGitHub, requireFile, python3, pkg-config
, audiofile, SDL2, hexdump, makeFlags ? []
, baseRom ?
  requireFile {
    name = "baserom.us.z64";
    message = ''
    This nix expression requires that baserom.us.z64 is
    already part of the store. To get this file you can dump your US Super Mario 64 cartridge's contents
    and add it to the nix store with nix-store --add-fixed sha256 <FILE>.
  '';
    sha256 = "17ce077343c6133f8c9f2d6d6d9a4ab62c8cd2aa57c40aea1f490b4c8bb21d91";
  }
}:

stdenv.mkDerivation rec {
  ...
  inherit makeFlags baseRom;
  ...
}
```

**Comment by IvarWithoutBones** [comment]
> Regarding licensing, sm64pc depends on n64-fast32-engine, which has a license stating that the source can be redistributed, but not the binary. This would definitely have to be marked as unfree.

Ah, okay. Switching to that :+1: 

> To avoid the infinite recursion, you could rename the argument to something other than baseRom, but you could also just directly set the default:

Thank you! learning new things every day haha. I've just pushed a WIP thing that makes use of that, and this works fine except for 1 problem: I can't replace the hash of `baseRom` based on the region being used. In other languages that'd be a simple if statement, but that doesn't seem to do the trick here. Input would be much appreciated

**Comment by MetaDark** [comment]
What was the issue you were having with if statements? It should have worked. Although, it might be easier to use a map for this case:

```nix
{
  sha256 = {
    "eu" = "0000000000000000000000000000000000000000000000000000000000000000";
    "jp" = "0000000000000000000000000000000000000000000000000000000000000000";
    "us" = "17ce077343c6133f8c9f2d6d6d9a4ab62c8cd2aa57c40aea1f490b4c8bb21d91";
  }.${region};
}
```

File: pkgs/games/sm64pc/default.nix
Comment (MetaDark) [inline pkgs/games/sm64pc/default.nix:29]: You shouldn't need to inherit `compileFlags` if it's not being used in any of the phases. That way you can still use `makeFlags` withing running into infinite recursion.

Comment (IvarWithoutBones) [reply pkgs/games/sm64pc/default.nix:29]: Ah, i thought i had to inherit it when i wanted to use it. I'll remove that.

**Comment by IvarWithoutBones** [comment]
> What was the issue you were having with if statements? It should have worked.

I'm not quite sure, there weren't any errors, it kinda just did not work. Either way, your solution indeed seems a lot cleaner, thanks again! Pushed a fixed version, with all the valid hashes according to members of their discord server.

File: pkgs/games/sm64pc/default.nix
Comment (MetaDark) [inline pkgs/games/sm64pc/default.nix:38]: Oh, one more thing. You can use `patchShebangs extract_assets.py` or `patchShebangs .` for easier maintenance, but that's up to you. Every thing else looks good!

**Review by kira-bruneau — APPROVED** [review]

File: pkgs/games/sm64pc/default.nix
Comment (IvarWithoutBones) [reply pkgs/games/sm64pc/default.nix:38]: Was not aware of that. Indeed a lot cleaner, I'll do that :+1:

**Review by kira-bruneau — APPROVED** [review]

File: pkgs/games/sm64pc/default.nix
Comment (TheBrainScrambler) [inline pkgs/games/sm64pc/default.nix:35]: You could add the BETTERCAMERA=1 makeFlag, it allows to control the camera with the mouse.
```suggestion
  makeFlags = [ "VERSION=${region}" "BETTERCAMERA=1" ] ++ compileFlags;
```

**Review by TheBrainScrambler — CHANGES_REQUESTED** [review]
Hello, I tried to build and run it with nixpkgs-review, and it works ! I just noticed that you could add the BETTERCAMERA makeFlag, or at least propose some option to add it.

File: pkgs/games/sm64pc/default.nix
Comment (MetaDark) [reply pkgs/games/sm64pc/default.nix:35]: I prefer using the better camera, but it probably shouldn't be the default. The `BETTERCAMERA` flag can be configured with the `compileFlags` argument.

Comment (IvarWithoutBones) [reply pkgs/games/sm64pc/default.nix:35]: Yep. That's why i added the `compileFlags` option in the first place. Personally i like the original camera better, but either way subjective patches different than the original game shouldn't be in the derivation by default IMO

Comment (TheBrainScrambler) [reply pkgs/games/sm64pc/default.nix:35]: I know they did in the AUR. The thing is, it only adds an option in the config to enable it, but doesn't enforce it if it is disabled, so I thought it was ok.
Also something else, I noticed that a sm64pc.config file spawned in my working directory by launching the game. It would be nice to handle its config later in some NixOS module.

Comment (IvarWithoutBones) [reply pkgs/games/sm64pc/default.nix:35]: > I know they did in the AUR. The thing is, it only adds an option in the config to enable it, but doesn't enforce it if it is disabled, so I thought it was ok.

Your suggestion just enables it though, and for subjective patches i don't think that's a good idea. There's already an option to enable it, with the `compileFlags` attribute. Though i guess that should be documented better, I'll add it to the `longDescription`.

> Also something else, I noticed that a sm64pc.config file spawned in my working directory by launching the game. It would be nice to handle its config later in some NixOS module.

Good idea! Though im currently not familiar enough with NixOS modules to do that :sweat_smile:

Comment (TheBrainScrambler) [reply pkgs/games/sm64pc/default.nix:35]: Yes, documenting it is probably the best solution.

**Review by kira-bruneau — APPROVED** [review]

**Review by TheBrainScrambler — APPROVED** [review]

**Comment by IvarWithoutBones** [comment]
Added darwin support, as i realized upstream has this. I do not own a Darwin machine to test, but in theory this should work.

**Comment by IvarWithoutBones** [comment]
It appears upstream got moved to `sm64pc/sm64ex`, so I've updated the derivation accordingly + updated to the latest git rev.

**Comment by IvarWithoutBones** [comment]
Anything blocking this from getting merged?

**Review by kira-bruneau — APPROVED** [review]
Looks, good to me! Unfortunately I don't have merge permissions.

**Pull request merged**
